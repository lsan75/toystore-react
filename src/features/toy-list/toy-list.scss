.ToyListComponent {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  padding: 1rem 1rem 0 0;

  align-content: stretch;
}

.ToyList__unselect {
  cursor: pointer;
  padding: 0 0 0 1rem;
  &.mdi:before {
    padding-right: 0.5rem;
  }
}
