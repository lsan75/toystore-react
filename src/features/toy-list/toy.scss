.ToyComponent {
  background: #eee;
  cursor: pointer;
  display: inline-block;
  height: 7rem;
  margin: 0 0 1rem 1rem;
  position: relative;
  transition: background-color 150ms ease-in-out;
  width: 7rem;

  &.selected, &.selected:hover {
    background: rgb(235, 185, 49);
  }

  &:hover {
    background: #fff;
  }
}

  .ToyComponent__image {
    font-size: 3rem;
    height: 100%;
    padding-top: 1.7rem;
    position: absolute;
    left: 0;
    text-align: center;
    top: 0;
    width: 100%;
  }
  .ToyComponent__text {
    position: absolute;
    padding: 0 0.4rem;
    bottom: 0.4rem;
    left: 0;
    font-size: 0.7rem;
    text-align: center;
    text-transform: uppercase;
    width: 100%;
  }