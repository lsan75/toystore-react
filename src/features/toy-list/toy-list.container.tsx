import * as React from 'react';
import { ToyListComponent } from './toy-list';
import { Toy } from './toy';
import { connect, Dispatch } from 'react-redux';

interface Props {
  toyList: Toy[];
  select: (toy: Toy) => {};
}

export class ToyListContainer extends React.Component<Props> {

  constructor(props: Props) {
    super(props);
  }

  render() {
    return (
      <ToyListComponent
        select={this.props.select}
        toyList={this.props.toyList}
      />
    );
  }

}

function mapStateToProps(state: Store) {
  return {
    toyList: state.toys.toyList
  };
}

function mapDispatchToProps(dispatch: Dispatch<Props>) {
  return {
    select: () => dispatch(selectToy(toy: Toy))
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ToyListContainer);