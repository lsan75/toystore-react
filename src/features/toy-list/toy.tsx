import * as React from 'react';
import '../../../node_modules/@mdi/font/css/materialdesignicons.min.css';
import './toy.css';

export interface Toy {
  title: string;
  icon: string;
  price: number;
  selected?: boolean;
}

interface Props {
  toy: Toy;
  select: (toy: Toy) => {};
}

export function ToyComponent(props: Props) {

  const classname = `
    ToyComponent
    mdi mdi-${props.toy.icon}
    ${props.toy.selected ? 'selected' : ''}
  `;

  return (
    <article
      className={classname}
      onClick={() => props.select(props.toy)}
    />
  );

}
