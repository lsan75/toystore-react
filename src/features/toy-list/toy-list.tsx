import * as React from 'react';
import { ToyComponent, Toy } from './toy';
import './toy-list.css';
import '../../../node_modules/@mdi/font/css/materialdesignicons.min.css';

interface Props {
  toyList: Toy[];
  select: (toy: Toy) => {};
  unselect: () => {};
}

interface State {
  hidden: boolean;
}

export class ToyListComponent extends React.Component<Props, State> {

  constructor(props: Props) {
    super(props);
    this.state = { hidden: false };
  }

  public content = () => this.props.toyList.map((toy: Toy) => {
    return <ToyComponent key={toy.title} toy={toy} select={this.props.select} />;
  })

  public resize = () => {
    return this.state.hidden ? ({ flexGrow: 0 }) : ({ flexGrow: 1 });
  }

  render() {
    return (
      <main className="ToyListContainer">
        <h1 style={{'paddingLeft': '1rem'}} onClick={() => this.setState({ hidden: !this.state.hidden})}>ToyList</h1>
        <section className="ToyListComponent" style={this.resize()}>{this.content()}</section>
        <section className="ToyList__unselect mdi mdi-close-circle" onClick={this.props.unselect}>Unselect All</section>
      </main>
    );
  }

}