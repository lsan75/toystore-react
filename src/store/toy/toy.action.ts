import { Dispatch } from 'react-redux';
import axios, { AxiosResponse } from 'axios';

import { ToyAction } from './toy.reducer';
import { Toy } from '../../features/toy-list/toy';

export const TOYS = {
  GET_TOYS: 'TOYS_GET_TOYS',
  SELECT_TOY: 'TOYS_SELECT_TOY',
  UNSELECT_ALL: 'TOYS_UNSELECT_ALL'
};

export function getToysAction() {
  return (dispatch: Dispatch<ToyAction>, getState: Function) => {

    const store = getState();
    if (store.toyReducer.toyList.length) { return; }

    axios.get('/toys').then((value: AxiosResponse<Toy[]>) => {
      dispatch(loadToysAction(value.data));
    }).catch(err => window.console.log(err));
  };
}

export function loadToysAction(toyList: Toy[]) {
  return {
    type: TOYS.GET_TOYS,
    toyList
  };
}

export function selectToyAction(toy: Toy): ToyAction {
  return {
    type: TOYS.SELECT_TOY,
    toy
  };
}

export function unselectAllAction(): ToyAction {
  return {
    type: TOYS.UNSELECT_ALL
  };
}