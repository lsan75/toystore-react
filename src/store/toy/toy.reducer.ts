import { Action } from 'redux';
import { Toy } from '../../features/toy-list/toy';
import { TOYS } from './toy.action';

export interface ToyReducerState {
  toyList: Toy[];
  counter: number;
}

export interface ToyAction extends Action {
  toyList?: Toy[];
  toy?: Toy;
}

export const defaultToyReducer: ToyReducerState = {
  toyList: [],
  counter: 0
};

export function toyReducer(state: ToyReducerState = defaultToyReducer, action: ToyAction) {

  switch (action.type) {

    case TOYS.GET_TOYS:
      return { ...state, toyList: action.toyList };

    case TOYS.SELECT_TOY:
      const newState = Object.assign({}, state);

      newState.toyList = newState.toyList.map(toy => {
        if (toy.title === (action.toy ? action.toy.title : null)) {
          toy.selected = !toy.selected;
        }
        return toy;
      });

      const counter = newState.toyList.filter(item => item.selected).length;

      return Object.assign({}, state, newState, { counter });

    default:
      return state;

  }
}
