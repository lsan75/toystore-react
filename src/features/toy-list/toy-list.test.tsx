import * as React from 'react'
import { configure, mount, ReactWrapper } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'
import { mountToJson } from 'enzyme-to-json'

import { ToyListComponent, Props, State } from './toy-list'
import { Toy } from './toy'

describe('ToyListComponent', () => {

  configure({ adapter: new Adapter() })
  let reactOutput: ReactWrapper
  let onSelect = jest.fn()
  let onUnselect = jest.fn()

  beforeEach(() => {
    const toyList: Toy[] = [
      {
        title: 'hop',
        icon: 'balloon',
        price: 10,
        selected: true
      }
    ]

    reactOutput = mount<Props, State>(
      <ToyListComponent
        toyList={toyList}
        select={onSelect}
        unselect={onUnselect}
      />
    )

  })

  afterEach(() => {
    reactOutput.unmount()
  })

  it('should render component', () => {
    expect(mountToJson(reactOutput)).toMatchSnapshot()
  })

  it('should toggle the list', () => {
    const h1 = reactOutput.find('h1')
    h1.simulate('click')

    expect(reactOutput.state('open')).toBe(false)
  })

  it('should unselect all', () => {
    const link = reactOutput.find('.ToyList__unselect')
    link.simulate('click')

    expect(onUnselect).toBeCalled()
  })

  it('should select a toy', () => {
    const toy = reactOutput.find('ToyComponent')
    toy.simulate('click')

    expect(onSelect).toBeCalledWith({
      title: 'hop',
      icon: 'balloon',
      price: 10,
      selected: true
    })
  })
})