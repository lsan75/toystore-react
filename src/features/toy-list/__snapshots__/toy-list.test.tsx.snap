// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToyListComponent should render component 1`] = `
<ToyListComponent
  select={[MockFunction]}
  toyList={
    Array [
      Object {
        "icon": "balloon",
        "price": 10,
        "selected": true,
        "title": "hop",
      },
    ]
  }
  unselect={[MockFunction]}
>
  <main>
    <h1
      onClick={[Function]}
      style={
        Object {
          "paddingLeft": "1rem",
        }
      }
    >
      ToyList
    </h1>
    <AnimateToggle
      duration={300}
      open={true}
    >
      <section
        style={
          Object {
            "height": "auto",
            "overflow": "hidden",
            "transition": "",
          }
        }
      >
        <div>
          <section
            className="ToyListComponent"
          >
            <ToyComponent
              key="hop"
              select={[MockFunction]}
              toy={
                Object {
                  "icon": "balloon",
                  "price": 10,
                  "selected": true,
                  "title": "hop",
                }
              }
            >
              <section
                className="
    ToyComponent
    selected
  "
                onClick={[Function]}
              >
                <article
                  className="
    ToyComponent__image
    mdi mdi-balloon
  "
                />
                <span
                  className="ToyComponent__text"
                >
                  hop
                </span>
              </section>
            </ToyComponent>
          </section>
        </div>
      </section>
    </AnimateToggle>
    <section
      className="ToyList__unselect mdi mdi-close-circle"
      onClick={[MockFunction]}
    >
      Unselect All
    </section>
  </main>
</ToyListComponent>
`;
